README
Scrub Quick Start

For those in a rush, here's how to get scrub working.
Download the latest release tarball and do the following to build scrub:

$ tar xvf scrub-2.6.0.tar.gz
scrub-2.6.0/
scrub-2.6.0/NEWS
scrub-2.6.0/test/
scrub-2.6.0/test/t02
scrub-2.6.0/test/t10.exp
scrub-2.6.0/test/t07.exp
scrub-2.6.0/test/t13.exp
scrub-2.6.0/test/t15.exp
scrub-2.6.0/test/t00.exp
scrub-2.6.0/test/pat.c
scrub-2.6.0/test/t05.exp
scrub-2.6.0/test/t18.exp
scrub-2.6.0/test/t17.exp
scrub-2.6.0/test/t20.exp
scrub-2.6.0/test/t03
scrub-2.6.0/test/t04.exp
scrub-2.6.0/test/t22.exp
scrub-2.6.0/test/t02.exp
scrub-2.6.0/test/t08
scrub-2.6.0/test/t06.exp
scrub-2.6.0/test/t04
scrub-2.6.0/test/t14.exp
scrub-2.6.0/test/t11.exp
scrub-2.6.0/test/README
scrub-2.6.0/test/Makefile.am
scrub-2.6.0/test/t09
scrub-2.6.0/test/tsig.c
scrub-2.6.0/test/t21.exp
scrub-2.6.0/test/t01.exp
scrub-2.6.0/test/t03.exp
scrub-2.6.0/test/t12
scrub-2.6.0/test/t16
scrub-2.6.0/test/t07
scrub-2.6.0/test/t16.exp
scrub-2.6.0/test/tgetsize.c
scrub-2.6.0/test/tsize.c
scrub-2.6.0/test/t15
scrub-2.6.0/test/t19.exp
scrub-2.6.0/test/tprogress.c
scrub-2.6.0/test/t17
scrub-2.6.0/test/t20
scrub-2.6.0/test/Makefile.in
scrub-2.6.0/test/t21
scrub-2.6.0/test/trand.c
scrub-2.6.0/test/t00
scrub-2.6.0/test/t18
scrub-2.6.0/test/t01
scrub-2.6.0/test/t12.exp
scrub-2.6.0/test/t19
scrub-2.6.0/test/t08.exp
scrub-2.6.0/test/t13
scrub-2.6.0/test/t22
scrub-2.6.0/test/t14
scrub-2.6.0/test/pad.c
scrub-2.6.0/test/t11
scrub-2.6.0/test/t09.exp
scrub-2.6.0/test/t05
scrub-2.6.0/test/aestest.c
scrub-2.6.0/test/t10
scrub-2.6.0/test/t06
scrub-2.6.0/scrub.spec
scrub-2.6.0/AUTHORS
scrub-2.6.0/aclocal.m4
scrub-2.6.0/ChangeLog
scrub-2.6.0/README
scrub-2.6.0/configure.ac
scrub-2.6.0/Makefile.am
scrub-2.6.0/configure
scrub-2.6.0/src/
scrub-2.6.0/src/hwrand.h
scrub-2.6.0/src/hwrand.c
scrub-2.6.0/src/sig.c
scrub-2.6.0/src/pattern.h
scrub-2.6.0/src/genrand.h
scrub-2.6.0/src/util.h
scrub-2.6.0/src/fillfile.c
scrub-2.6.0/src/getsize.c
scrub-2.6.0/src/Makefile.am
scrub-2.6.0/src/scrub.c
scrub-2.6.0/src/aes.h
scrub-2.6.0/src/sig.h
scrub-2.6.0/src/getsize.h
scrub-2.6.0/src/util.c
scrub-2.6.0/src/fillfile.h
scrub-2.6.0/src/filldentry.h
scrub-2.6.0/src/Makefile.in
scrub-2.6.0/src/pattern.c
scrub-2.6.0/src/progress.c
scrub-2.6.0/src/filldentry.c
scrub-2.6.0/src/progress.h
scrub-2.6.0/src/aes.c
scrub-2.6.0/src/genrand.c
scrub-2.6.0/man/
scrub-2.6.0/man/Makefile.am
scrub-2.6.0/man/scrub.1
scrub-2.6.0/man/scrub.1.in
scrub-2.6.0/man/Makefile.in
scrub-2.6.0/scrub.spec.in
scrub-2.6.0/INSTALL
scrub-2.6.0/Makefile.in
scrub-2.6.0/META
scrub-2.6.0/COPYING
scrub-2.6.0/DISCLAIMER
scrub-2.6.0/libscrub/
scrub-2.6.0/libscrub/libscrub.map
scrub-2.6.0/libscrub/scrub.h
scrub-2.6.0/libscrub/README
scrub-2.6.0/libscrub/Makefile.am
scrub-2.6.0/libscrub/Makefile.in
scrub-2.6.0/libscrub/libscrub.c
scrub-2.6.0/libscrub/libscrub.pc.in
scrub-2.6.0/libscrub/tscrub.c
scrub-2.6.0/config/
scrub-2.6.0/config/ac_pkgconfig.m4
scrub-2.6.0/config/ltmain.sh
scrub-2.6.0/config/config.h.in
scrub-2.6.0/config/depcomp
scrub-2.6.0/config/config.sub
scrub-2.6.0/config/x_ac_expand_install_dirs.m4
scrub-2.6.0/config/x_ac_check_pthreads.m4
scrub-2.6.0/config/x_ac_meta.m4
scrub-2.6.0/config/config.guess
scrub-2.6.0/config/install-sh
scrub-2.6.0/config/missing
$
$ cd scrub-2.6.0
$ ./configure
checking build system type... x86_64-unknown-linux-gnu
checking host system type... x86_64-unknown-linux-gnu
checking target system type... x86_64-unknown-linux-gnu
checking metadata... yes
checking installation directory variables... yes
checking for a BSD-compatible install... /usr/bin/install -c
checking whether build environment is sane... yes
checking for a thread-safe mkdir -p... /bin/mkdir -p
checking for gawk... gawk
checking whether make sets $(MAKE)... yes
checking whether to enable maintainer-specific portions of Makefiles... no
checking for gcc... gcc
checking for C compiler default output file name... a.out
checking whether the C compiler works... yes
checking whether we are cross compiling... no
checking for suffix of executables... 
checking for suffix of object files... o
checking whether we are using the GNU C compiler... yes
checking whether gcc accepts -g... yes
checking for gcc option to accept ISO C89... none needed
checking for style of include used by make... GNU
checking dependency style of gcc... gcc3
checking how to run the C preprocessor... gcc -E
checking for grep that handles long lines and -e... /bin/grep
checking for egrep... /bin/grep -E
checking for ANSI C header files... yes
checking for sys/types.h... yes
checking for sys/stat.h... yes
checking for stdlib.h... yes
checking for string.h... yes
checking for memory.h... yes
checking for strings.h... yes
checking for inttypes.h... yes
checking for stdint.h... yes
checking for unistd.h... yes
checking minix/config.h usability... no
checking minix/config.h presence... no
checking for minix/config.h... no
checking whether it is safe to define __EXTENSIONS__... yes
checking for ANSI C header files... (cached) yes
checking getopt.h usability... yes
checking getopt.h presence... yes
checking for getopt.h... yes
checking stdbool.h usability... yes
checking stdbool.h presence... yes
checking for stdbool.h... yes
checking pthread.h usability... yes
checking pthread.h presence... yes
checking for pthread.h... yes
checking for a sed that does not truncate output... /bin/sed
checking for fgrep... /bin/grep -F
checking for ld used by gcc... /usr/bin/ld
checking if the linker (/usr/bin/ld) is GNU ld... yes
checking for BSD- or MS-compatible name lister (nm)... /usr/bin/nm -B
checking the name lister (/usr/bin/nm -B) interface... BSD nm
checking whether ln -s works... yes
checking the maximum length of command line arguments... 1966080
checking whether the shell understands some XSI constructs... yes
checking whether the shell understands "+="... yes
checking for /usr/bin/ld option to reload object files... -r
checking for objdump... objdump
checking how to recognize dependent libraries... pass_all
checking for ar... ar
checking for strip... strip
checking for ranlib... ranlib
checking command to parse /usr/bin/nm -B output from gcc object... ok
checking for dlfcn.h... yes
checking for objdir... .libs
checking if gcc supports -fno-rtti -fno-exceptions... no
checking for gcc option to produce PIC... -fPIC -DPIC
checking if gcc PIC flag -fPIC -DPIC works... yes
checking if gcc static flag -static works... yes
checking if gcc supports -c -o file.o... yes
checking if gcc supports -c -o file.o... (cached) yes
checking whether the gcc linker (/usr/bin/ld -m elf_x86_64) supports shared libraries... yes
checking whether -lc should be explicitly linked in... no
checking dynamic linker characteristics... GNU/Linux ld.so
checking how to hardcode library paths into programs... immediate
checking whether stripping libraries is possible... yes
checking if libtool supports shared libraries... yes
checking whether to build shared libraries... yes
checking whether to build static libraries... yes
checking whether to install pkg-config *.pc files... ${libdir}/pkgconfig
checking whether byte ordering is bigendian... no
checking for an ANSI C-conforming const... yes
checking for getopt_long... yes
checking for posix_memalign... yes
checking for memalign... yes
checking for posix_fadvise... yes
checking for rand_r... yes
checking for random_r... yes
checking how to link against pthreads... -lpthread
checking for special C compiler options needed for large files... no
checking for _FILE_OFFSET_BITS value needed for large files... no
configure: creating ./config.status
config.status: creating Makefile
config.status: creating scrub.spec
config.status: creating src/Makefile
config.status: creating man/Makefile
config.status: creating man/scrub.1
config.status: creating test/Makefile
config.status: creating libscrub/Makefile
config.status: creating libscrub/libscrub.pc
config.status: creating config/config.h
config.status: executing depfiles commands
config.status: executing libtool commands
$
$ make
gcc -DHAVE_CONFIG_H -I. -I../config     -g -O2 -MT aes.o -MD -MP -MF .deps/aes.Tpo -c -o aes.o aes.c
mv -f .deps/aes.Tpo .deps/aes.Po
gcc -DHAVE_CONFIG_H -I. -I../config     -g -O2 -MT filldentry.o -MD -MP -MF .deps/filldentry.Tpo -c -o filldentry.o filldentry.c
mv -f .deps/filldentry.Tpo .deps/filldentry.Po
gcc -DHAVE_CONFIG_H -I. -I../config     -g -O2 -MT fillfile.o -MD -MP -MF .deps/fillfile.Tpo -c -o fillfile.o fillfile.c
mv -f .deps/fillfile.Tpo .deps/fillfile.Po
gcc -DHAVE_CONFIG_H -I. -I../config     -g -O2 -MT genrand.o -MD -MP -MF .deps/genrand.Tpo -c -o genrand.o genrand.c
mv -f .deps/genrand.Tpo .deps/genrand.Po
gcc -DHAVE_CONFIG_H -I. -I../config     -g -O2 -MT getsize.o -MD -MP -MF .deps/getsize.Tpo -c -o getsize.o getsize.c
mv -f .deps/getsize.Tpo .deps/getsize.Po
gcc -DHAVE_CONFIG_H -I. -I../config     -g -O2 -MT hwrand.o -MD -MP -MF .deps/hwrand.Tpo -c -o hwrand.o hwrand.c
mv -f .deps/hwrand.Tpo .deps/hwrand.Po
gcc -DHAVE_CONFIG_H -I. -I../config     -g -O2 -MT pattern.o -MD -MP -MF .deps/pattern.Tpo -c -o pattern.o pattern.c
mv -f .deps/pattern.Tpo .deps/pattern.Po
gcc -DHAVE_CONFIG_H -I. -I../config     -g -O2 -MT progress.o -MD -MP -MF .deps/progress.Tpo -c -o progress.o progress.c
mv -f .deps/progress.Tpo .deps/progress.Po
gcc -DHAVE_CONFIG_H -I. -I../config     -g -O2 -MT scrub.o -MD -MP -MF .deps/scrub.Tpo -c -o scrub.o scrub.c
mv -f .deps/scrub.Tpo .deps/scrub.Po
gcc -DHAVE_CONFIG_H -I. -I../config     -g -O2 -MT sig.o -MD -MP -MF .deps/sig.Tpo -c -o sig.o sig.c
mv -f .deps/sig.Tpo .deps/sig.Po
gcc -DHAVE_CONFIG_H -I. -I../config     -g -O2 -MT util.o -MD -MP -MF .deps/util.Tpo -c -o util.o util.c
mv -f .deps/util.Tpo .deps/util.Po
/bin/sh ../libtool --tag=CC   --mode=link gcc  -g -O2   -o scrub aes.o filldentry.o fillfile.o genrand.o getsize.o hwrand.o pattern.o progress.o scrub.o sig.o util.o -lpthread 
libtool: link: gcc -g -O2 -o scrub aes.o filldentry.o fillfile.o genrand.o getsize.o hwrand.o pattern.o progress.o scrub.o sig.o util.o  -lpthread
make[1]: Leaving directory `/home/garlick/proj/scrub/scrub/scrub-2.6.0/src'
Making all in man
make[1]: Entering directory `/home/garlick/proj/scrub/scrub/scrub-2.6.0/man'
make[1]: Nothing to be done for `all'.
make[1]: Leaving directory `/home/garlick/proj/scrub/scrub/scrub-2.6.0/man'
Making all in test
make[1]: Entering directory `/home/garlick/proj/scrub/scrub/scrub-2.6.0/test'
make[1]: Nothing to be done for `all'.
make[1]: Leaving directory `/home/garlick/proj/scrub/scrub/scrub-2.6.0/test'
make[1]: Entering directory `/home/garlick/proj/scrub/scrub/scrub-2.6.0'
make[1]: Nothing to be done for `all-am'.
make[1]: Leaving directory `/home/garlick/proj/scrub/scrub/scrub-2.6.0'
$

To scrub the second partition on Linux disk /dev/sda using default NNSA
algorithm, do the following.  Note that for normal SATA disks this will
take some time.

$ src/scrub /dev/sda2
scrub: using NNSA NAP-14.1-C patterns
scrub: please verify that device size below is correct!
scrub: scrubbing /dev/sda2 6234762240 bytes (~5GB)
scrub: random  |................................................|
scrub: random  |................................................|
scrub: 0x00    |................................................|
scrub: verify  |................................................|
$


NEWS
========================================================================
scrub-2.6.1 Release Notes                                     2014-08-26
========================================================================

* Add dist tag to release (TOSS-2665)

* OS X build/test failrues (issue 31)

* Minor autotools fixes.

========================================================================
scrub-2.6.0 Release Notes                                     2014-08-20
========================================================================

* Added support for multiple file/device arguments [Peter Hyman]

* Added --dry-run option

* Handle incorrect octal esc in custom pattern (issue 28)

* Dont truncate custom pattern in display (issue 27)

* Add 'verify' pattern:  Quick Fill with 0x00 and verify (issue 22)

* Add unit test coverage for -X, --dry-run, multiple files.

* Untrack and ignore generated files in git (issue 15)

* Project repository:  migrated source/releases to github, restored
  pre-v2.2 history from sourceforge;  retain Google Code issue tracker
  and wiki for now.

* Doc update: README.md, tests/README, AUTHORS, source code headers.

========================================================================
scrub-2.5.3 Release Notes                                     2014-05-20
========================================================================

* AIX 6.1: handle large SCSI disk,
  ignore EINVAL from fsync [Tim Starrin] (issue 26)

* OS X ML/Darwin: fix build error.

* Newer gcc: fix /dev/urandom build error.

========================================================================
scrub-2.5.2 Release Notes                                     2012-06-20
========================================================================

* Fix scrub abort on files/disks > 2G in size (issue 19)
  This was introduced in 2.5.0 and the failure is quick and obvious.

* Fix loss of precision in disk size approximation (issue 18)

* Add -t option to disable creation of thread to generate randomness
  in parallel with I/O.

* Prepare source for a libscrub api (work in progress)

========================================================================
scrub-2.5.1 Release Notes                                     2012-05-08
========================================================================

* Cosmetic changes to man page and help output.

========================================================================
scrub-2.5.0 Release Notes                                     2012-05-03
========================================================================

* Generate randomness and do I/O in parallel [Richard B. Hill]

* Replace O_SYNC with O_DIRECT+fsync+posix_fadvise [H. Peter Anvin]
  This produces a significant performance improvement, at least on Linux.

* Add support for RDRAND instr. [H. Peter Anvin]

* Use <stdbool.h> if the system has it [H. Peter Anvin]

* Fix incorrect assertion in scrub --dirent handling (issue 9) [Peter Hyman]

* Clean up tmp files after scrub --freespace (issue 9) [Peter Hyman]

* Allow scrub --pattern=custom="string" (issue 11) [Peter Hyman]
  Allow C style numerical escapes in string: \nnn (octal) and \xnn (hex).
  This is a single-pass sequence with a custom pattern.

* Change default buffer size from 1m to 4m [H. Peter Anvin]

* Avoid closing and re-opening /dev/urandom [H. Peter Anvin]

========================================================================
scrub-2.4.2 Release Notes                                     2011-08-29
========================================================================

Same as 2.4.1 but check in missing spec file (as opposed to spec.in).

========================================================================
scrub-2.4.1 Release Notes                                     2011-08-29
========================================================================

* Fix bug in pfitzner7 pattern def and add fillff, fillzero
  methods (issue 8) [Peter Hyman]
    
* Change dod method to perform random pass first and skip extra pass.

* Add --help option that lists available scrub methods.

* Make dist tarball work with rpmbuild -ta.

========================================================================
scrub-2.4 Release Notes                                       2010-11-03
========================================================================

* Added usarmy, pfitzner7, and pfitzner33 methods.

========================================================================
scrub-2.3 Release Notes                                       2010-05-14
========================================================================

* Added -L option to not follow symbolic links [Timothy Boronczyk]

* Treat a zero return from write(2) as an error.

* New patterns:
  - random2: two random passes
  - schneier (one 0x00 pass, one 0xff pass, and five random passes) [Joe]

========================================================================
scrub-2.2 Release Notes                                       2009-07-29
========================================================================

* Add --pattern random (one random pass) mode.

* Portability fixes for Slackware [Timothy Boronczyk]

* Fix bug where scrub --remove on a zero length file would abort
  without removing the file.  Issue a warning and remove it.

========================================================================
scrub-2.1 Release Notes                                       2008-11-14
========================================================================

* Fix a bug where some data from the original file could appear in
  the first disk block after a verified scrub when running without 
  the -S (--no-signature) option.

========================================================================
scrub-2.0 Release Notes                                       2008-05-30
========================================================================

* Longopts support.


scrub.c
/*****************************************************************************\
 *  $Id: scrub.c 78 2006-02-15 01:05:03Z garlick $
 *****************************************************************************
 *  Copyright (C) 2001-2002 The Regents of the University of California.
 *  Produced at Lawrence Livermore National Laboratory (cf, DISCLAIMER).
 *  Written by Jim Garlick <garlick@llnl.gov>.
 *  UCRL-CODE-2003-006.
 *  
 *  This file is part of Scrub, a program for erasing disks.
 *  For details, see <http://www.llnl.gov/linux/scrub/>.
 *  
 *  Scrub is free software; you can redistribute it and/or modify it under
 *  the terms of the GNU General Public License as published by the Free
 *  Software Foundation; either version 2 of the License, or (at your option)
 *  any later version.
 *  
 *  Scrub is distributed in the hope that it will be useful, but WITHOUT ANY
 *  WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 *  FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
 *  details.
 *  
 *  You should have received a copy of the GNU General Public License along
 *  with Scrub; if not, write to the Free Software Foundation, Inc.,
 *  59 Temple Place, Suite 330, Boston, MA  02111-1307  USA.
\*****************************************************************************/

/* Scrub a raw disk or plain file.
 */

#if defined(linux) || defined(sun) || defined(UNIXWARE) || defined(__hpux)
#define _LARGEFILE_SOURCE 
#define _FILE_OFFSET_BITS 64
#endif

#if defined(_AIX)
#define _LARGE_FILES
#include <sys/mode.h>
#endif

#include <sys/stat.h>
#include <fcntl.h>
#include <stdio.h>
#include <libgen.h>
#include <string.h>
#include <unistd.h>
#include <stdlib.h>
#include <libgen.h>
#include <assert.h>
#include <sys/param.h> /* MAXPATHLEN */

#include "genrand.h"
#include "fillfile.h"
#include "filldentry.h"
#include "getsize.h"
#include "progress.h"
#include "util.h"
#include "sig.h"

#define RANDOM	0x0100
#define VERIFY  0x0200

static const int dirent_pattern[] = { 0x55, 0x22, 0x55, 0x22, 0x55, 0x22 };

static const int old_pattern[] = { 0, 0xff, 0xaa, RANDOM, 0x55, VERIFY };
static const int fastold_pattern[] = { 0, 0xff, 0xaa, 0x55, VERIFY };
static const int nnsa_pattern[] = { RANDOM, RANDOM, 0, VERIFY };
static const int dod_pattern[] = { 0, 0xff, RANDOM, 0, VERIFY };
static const int bsi_pattern[] = { 0xff, 0xfe, 0xfd, 0xfb, 0xf7,
                                   0xef, 0xdf, 0xbf, 0x7f };

typedef enum { false, true } bool;
typedef enum { NOEXIST, REGULAR, SPECIAL, OTHER } filetype_t;

static char *pat2str(int pat);
static void usage(void);
static off_t blkalign(off_t offset, int blocksize);
static filetype_t filetype(char *path);

/* NOTE: default blocksize was 8K in scrub 1.8, however on hpux
 * it was found that raising the default to 1M raised performance from
 * ~1.3 MB/s to 20MB/s. [Graham Smith]
 * Probably it won't hurt on the other OS's.
 */
#define BUFSIZE (1024*1024) /* default blocksize */

char *prog;

static void 
usage(void)
{
    fprintf(stderr, "Usage: %s [-v] [-f] [-p pat] [-b blocksize] [-X] [-D newname] [-r] file\n", prog);
    fprintf(stderr, "\t-p select scrub patterns (see scrub(1))\n");
    fprintf(stderr, "\t-b overrides default I/O buffer size of %d bytes\n", BUFSIZE);
    fprintf(stderr, "\t-X create file and keep writing until write fails, then scrub\n");
    fprintf(stderr, "\t-D after scrubbing the file, scrub the directory entry and rename\n");
    fprintf(stderr, "\t-f scrub even if file has signature from previous scrub\n");
    fprintf(stderr, "\t-r remove file after scrub\n");
    fprintf(stderr, "\t-v display scrub version and exit\n");

    exit(1);
}

static char *
pat2str(int pat)
{
    static char str[255];

    if (pat == RANDOM)
        snprintf(str, sizeof(str), "random");
    else if (pat == VERIFY)
        snprintf(str, sizeof(str), "verify");
    else
        snprintf(str, sizeof(str), "0x%x", pat);
    return str;
}

/* Round 'offset' up to an even multiple of 'blocksize'.
 */
static off_t
blkalign(off_t offset, int blocksize)
{
    off_t r = offset % blocksize;

    if (r > 0) 
        offset += (blocksize - r);

    return offset;
}

/* Return the type of file represented by 'path'.
 */
static filetype_t
filetype(char *path)
{
    struct stat sb;
    filetype_t res = NOEXIST;

    if (stat(path, &sb) == 0) {
        if (S_ISREG(sb.st_mode))
            res = REGULAR;
#if defined(linux)
        /* on modern Linux, disks only have a block interface */
        else if (S_ISCHR(sb.st_mode) || S_ISBLK(sb.st_mode))
#else
        else if (S_ISCHR(sb.st_mode))
#endif
            res = SPECIAL;
        else
            res = OTHER;
    }
    return res;
}

/* Scrub 'path', a file/device of size 'size'.
 * Fill with the bytes in 'pat', an array of 'npat' elements.
 * Use 'bufsize' length for I/O buffers.
 */
static void
scrub(char *path, off_t size, const int pat[], int npat, int bufsize, int Sopt)
{
    unsigned char *buf;
    int i;
    prog_t p;
    char sizestr[80];

    if (!(buf = malloc(bufsize))) {
        fprintf(stderr, "%s: out of memory\n", prog);
        exit(1);
    }

    size2str(sizestr, sizeof(sizestr), size);
    printf("%s: scrubbing %s %s\n", prog, path, sizestr);

    initrand();
    for (i = 0; i < npat; i++) {
        printf("%s: %-8s", prog, pat2str(pat[i]));
        progress_create(&p, 50);
        if (pat[i] == RANDOM) {
            churnrand();
            fillfile(path, size, buf, bufsize, 
                    (progress_t)progress_update, p, (refill_t)genrand);
        } else if (pat[i] == VERIFY) {
            assert(i > 0);
            assert(pat[i - 1] != RANDOM);
            assert(pat[i - 1] != VERIFY);
            /* depends on previous pass contents of buf being intact */
            checkfile(path, size, buf, bufsize, 
                    (progress_t)progress_update, p);
        } else {
            assert(pat[i] <= 0xff);
            memset(buf, pat[i], bufsize);
            fillfile(path, size, buf, bufsize, 
                    (progress_t)progress_update, p, NULL);
        }
        progress_destroy(p);
    }
    if (!Sopt)
        writesig(path, bufsize);

    free(buf);
}

/* Scrub free space (-X).
 */
static void
scrub_free(char *path, const int pat[], int npat, int bufsize, int Sopt)
{
    unsigned char *buf;
    off_t size; 

    assert(filetype(path) == NOEXIST || filetype(path) == REGULAR);

    /* special scrub for first pass */
    if (!(buf = malloc(bufsize))) {
        fprintf(stderr, "%s: out of memory\n", prog);
        exit(1);
    }
    assert(npat > 0);
    assert(pat[0] != VERIFY);
    printf("%s: filling file system by expanding %s\n", prog, path);
    printf("%s: %-8s...\n", prog, pat2str(pat[0]));
    fflush(stdout);
    /* XXX add some feedback */
    if (pat[0] == RANDOM) {
        churnrand();
        size = growfile(path, buf, bufsize, (refill_t)genrand);
    } else {
        memset(buf, pat[0], bufsize);
        size = growfile(path, buf, bufsize, NULL);
    }
    free(buf);

    /* remaining passes as usual */
    if (npat > 1)
        scrub(path, size, pat+1, npat-1, bufsize, Sopt);
}

/* Scrub name component of a directory entry through succesive renames.
 */
static void
scrub_dirent(char *path, char *newpath)
{
    const int *pat = dirent_pattern;
    int npat = sizeof(dirent_pattern)/sizeof(dirent_pattern[0]);
    prog_t p;
    int i;

    assert(filetype(path) == REGULAR);

    printf("%s: scrubbing directory entry\n", prog);

    for (i = 0; i < npat; i++) {
        assert(pat[i] != 0);
        assert(pat[i] != RANDOM);
        assert(pat[i] != VERIFY);
        printf("%s: %-8s", prog, pat2str(pat[i]));
        progress_create(&p, 50);
        filldentry(path, pat[i]); /* path: in/out */
        progress_update(p, 1.0);
        progress_destroy(p);
    }
    if (rename(path, newpath) < 0) {
        fprintf(stderr, "%s: error renaming %s to %s\n", prog, path, newpath);
        exit(1);
    }
}

/* Scrub a regular file.
 */
static void 
scrub_file(char *path, const int pat[], int npat, int bufsize, int Sopt)
{
    off_t size; 
    struct stat sb;

    assert(filetype(path) == REGULAR);

    if (stat(path, &sb) < 0) {
        fprintf(stderr, "%s: stat ", prog);
        perror(path);
        exit(1);
    }
    if (sb.st_size == 0) {
        fprintf(stderr, "%s: %s is zero length\n", prog, path);
        exit(1);
    }
    size = blkalign(sb.st_size, sb.st_blksize);
    if (size != sb.st_size) {
        printf("%s: padding %s with %d bytes to fill last fs block\n", 
                prog, path, (int)(size - sb.st_size)); 
    }
    scrub(path, size, pat, npat, bufsize, Sopt);
}

/* Scrub apple resource fork component of file.
 */
#if __APPLE__
static void
scrub_resfork(char *path, const int pat[], int npat, int bufsize)
{
    struct stat rsb;
    char rpath[MAXPATHLEN];
    off_t rsize; 

    assert(filetype(path) == REGULAR);
    (void)snprintf(rpath, sizeof(rpath), "%s/..namedfork/rsrc", path);
    if (stat(rpath, &rsb) < 0)
        return;
    if (rsb.st_size == 0) {
        printf("%s: skipping zero length resource fork: %s\n", prog, rpath);
        return;
    }
    printf("%s: scrubbing resource fork: %s\n", prog, rpath);
    rsize = blkalign(rsb.st_size, rsb.st_blksize);
    if (rsize != rsb.st_size) {
        printf("%s: padding %s with %d bytes to fill last fs block\n", 
                        prog, rpath, (int)(rsize - rsb.st_size)); 
    }
    scrub(rpath, rsize, pat, npat, bufsize, 0);
}
#endif

/* Scrub a special file corresponding to a disk.
 */
static void
scrub_disk(char *path, off_t size, const int pat[], int npat, int bufsize, int Sopt)
{
    assert(filetype(path) == SPECIAL);
    if (size == 0) {
        size = getsize(path);
        if (size == 0) {
            fprintf(stderr, "%s: could not determine size, use -s\n", prog);
            exit(1);
        }
        printf("%s: please verify that device size below is correct!\n", prog);
    }
    scrub(path, size, pat, npat, bufsize, Sopt);
}

int 
main(int argc, char *argv[])
{
    const int *pat = nnsa_pattern;
    int npat = sizeof(nnsa_pattern)/sizeof(nnsa_pattern[0]);
    bool Xopt = false;
    int bopt = BUFSIZE;
    off_t sopt = 0;
    char *Dopt = NULL;
    char *filename = NULL;
    int fopt = 0;
    int Sopt = 0;
    int ropt = 0;

    extern int optind;
    extern char *optarg;
    int c;

    /* Handle arguments.
     */
    prog = basename(argv[0]);
    while ((c = getopt(argc, argv, "p:D:Xb:s:fSrv")) != EOF) {
        switch (c) {
        case 'v':   /* Print version and exit */
            printf("scrub version %s\n", SCRUB_VERSION);
            exit(0);
        case 'p':   /* Override default pattern with dod|nnsa|old|fastold */
            if (!strcmp(optarg, "dod") || !strcmp(optarg, "DOD")) {
                pat = dod_pattern;
                npat = sizeof(dod_pattern)/sizeof(dod_pattern[0]);
            } else if (!strcmp(optarg, "nnsa") || !strcmp(optarg, "NNSA")) {
                pat = nnsa_pattern;
                npat = sizeof(nnsa_pattern)/sizeof(nnsa_pattern[0]);
            } else if (!strcmp(optarg, "old") || !strcmp(optarg, "OLD")) {
                pat = old_pattern;
                npat = sizeof(old_pattern)/sizeof(old_pattern[0]);
            } else if (!strcmp(optarg, "fastold") || !strcmp(optarg, "FASTOLD")) {
                pat = fastold_pattern;
                npat = sizeof(fastold_pattern)/sizeof(fastold_pattern[0]);
            } else if (!strcmp(optarg, "bsi") || !strcmp(optarg, "BSI")) {
                pat = bsi_pattern;
                npat = sizeof(bsi_pattern)/sizeof(bsi_pattern[0]);
            } else 
                usage();
            break;
        case 'X':   /* fill filesystem */
            Xopt = true;
            break;
        case 'D':   /* scrub name in dirent through successive renames */
            Dopt = optarg;
            break;
        case 'r':   /* remove file when done */
            ropt = 1;
            break;
        case 'b':   /* override blocksize */
            bopt = str2int(optarg);
            if (bopt == 0) {
                fprintf(stderr, "%s: error parsing blocksize string\n", prog);
                exit(1);
            }
            break;
        case 's':   /* override size of special file */
            sopt = str2size(optarg);
            if (sopt == 0) {
                fprintf(stderr, "%s: error parsing size string\n", prog);
                exit(1);
            }
            break;
        case 'f':   /* force scrub even if already done */
            fopt = 1;
            break;
        case 'S':   /* do not write scrub signature */
            Sopt = 1;
            break;
        default:
            usage();
        }
    }
    if (argc - optind != 1)
        usage();
    filename = argv[optind++];

    /* Verify arguments.
     */
    if (filename == NULL)
        usage();
    if (Xopt) {
        if (filetype(filename) == SPECIAL) {
            fprintf(stderr, "%s: -X cannot be used on special files\n", prog);
            exit(1);
        }
        if (sopt > 0) {
            fprintf(stderr, "%s: -s and -X cannot be used together\n", prog);
            exit(1);
        }
        if (Dopt) {
            fprintf(stderr, "%s: -D and -X cannot be used together\n", prog);
            exit(1);
        }
    } else {
        switch (filetype(filename)) {
            case NOEXIST:
                fprintf(stderr, "%s: %s does not exist\n", prog, filename);
                exit(1);
                break;
            case OTHER:
                fprintf(stderr, "%s: %s is not a reg file or %sdisk device\n", 
                    prog, filename,
#if defined(linux)
                    ""
#else
                    "raw "
#endif
                    );
                exit(1);
                break;
            case SPECIAL:
                if (Dopt) {
                    fprintf(stderr, "%s: cannot use -D with special file\n", prog);
                    exit(1);
                }
                if (ropt) {
                    fprintf(stderr, "%s: cannot use -r with special file\n", prog);
                    exit(1);
                }
                break;
            case REGULAR:
                if (sopt > 0) {
                    fprintf(stderr, "%s: cannot use -s with regular file\n", prog);
                    exit(1);
                }
                if (Dopt && *Dopt != '/' && *filename == '/') {
                    fprintf(stderr, "%s: %s should be a full path like %s\n",
                            prog, Dopt, filename);
                    exit(1);
                }
                break;
        }
        if (access(filename, R_OK|W_OK) < 0) {
            fprintf(stderr, "%s: no permission to scrub %s\n", prog, filename);
            exit(1);
        }
        if (checksig(filename, bopt) && !fopt) {
            fprintf(stderr, "%s: %s has already been scrubbed? (use -f to force)\n",
                        prog, filename);
            exit(1);
        }
    }


    if (sizeof(off_t) < 8) {
        fprintf(stderr, "%s: warning: not using 64 bit file offsets\n", prog);
    }

    /* Scrub.
     */
    printf("%s: using %s patterns\n", prog, 
            (pat == dod_pattern)        ? "DoD 5220.22-M" 
            : (pat == nnsa_pattern)     ? "NNSA NAP-14.x" 
            : (pat == old_pattern)      ? "pre v1.7 scrub" 
            : (pat == fastold_pattern)  ? "pre v1.7 scrub (skip random)" 
            : (pat == bsi_pattern)      ? "BSI pattern" 
            : "unknown pattern");

    if (Xopt) {                                     /* scrub free */
        scrub_free(filename, pat, npat, bopt, Sopt);

    } else if (filetype(filename) == REGULAR) {     /* scrub file */
        scrub_file(filename, pat, npat, bopt, Sopt);
#if __APPLE__
        scrub_resfork(filename, pat, npat, bopt);
#endif
        if (Dopt) { /* XXX destroys 'filename' */
            scrub_dirent(filename, Dopt);
            filename = Dopt; /* -r needs this below */
        }

    } else if (filetype(filename) == SPECIAL) {     /* scrub disk */  
        scrub_disk(filename, sopt, pat, npat, bopt, Sopt);
    }

    /* unlink file at the end */
    if (ropt && filetype(filename) == REGULAR) {
        printf("%s: unlinking %s\n", prog, filename);
        if (unlink(filename) != 0) {
            perror(filename);
            exit(1);
        }
    }

    exit(0);
}

/*
 * vi:tabstop=4 shiftwidth=4 expandtab
 */
